### PR TITLE
Add the ability to keep beans from application classes

### PR DESCRIPTION
**Pull request by geoand**
This provides a useful middle-ground between removing all unused beans and removing none.

It could be useful in cases such when one wants to use `@Inject BeanManager beanManager` and get all the beans of a specific type.

### CODE REVIEW

File: docs/src/main/asciidoc/cdi-reference.adoc [hunk 1]
```diff
@@ -274,4 +274,18 @@ An unused bean:
...
+This could be useful in cases where bean discovery is needed, like when using
+
+[source,java]
+----
+@Inject
+BeanManager beanManager;
+
+...
+
+beanManager.getBeans(SomeService.class)
```
Comment (stuartwdouglas) [inline docs/src/main/asciidoc/cdi-reference.adoc]: This is more of a question for @mkouba  than a comment on the PR, but if Instance<SomeService> was injected in this case would this be enough to keep the beans from being removed?

**Comment by n1hility** [comment]
Is the use case here just compatibility with old apps? I think we should be very careful about preventing DCE. Once we take that step it’s very hard to change.

File: docs/src/main/asciidoc/cdi-reference.adoc (hunk 1 above)
Comment (geoand) [reply docs/src/main/asciidoc/cdi-reference.adoc]: When I tried it, the manager got injected just fine but no beans were present

**Comment by geoand** [comment]
I didn't really have compatibility with old apps in mind. I was looking for a way to be able to inject all beans of a type (I was thinking of some sort of proxy bean that would delegate to the others)

**Comment by geoand** [comment]
I was also thinking that perhaps a BuildItem could be added that when present would be automatically activate this feature. This would be useful for things like #1301

File: docs/src/main/asciidoc/cdi-reference.adoc (hunk 1 above)
Comment (mkouba) [reply docs/src/main/asciidoc/cdi-reference.adoc]: `@Inject Instance<SomeService>` in app code would prevent the `SomeService` bean from being removed. `@Inject Instance<Object>` effectively disables the "removal" feature completely.

**Review by mkouba — COMMENTED** [review]
TBH I don't understand the use case. How would this help with https://github.com/quarkusio/quarkus/pull/1301?

**Comment by geoand** [comment]
> TBH I don't understand the use case. How would this help with #1301?

@mkouba When using Aesh, Quarkus can not know that the application code would be pulling beans from Arc, see: https://github.com/quarkusio/quarkus/blob/12ff0a17d9d899fab14ffa751c4ead44c8380471/integration-tests/command-line-apps/src/test/resources/projects/aesh/src/main/java/org/acme/CapitalizeCommand.java#L33

**Comment by mkouba** [comment]
Well, in that particular case I would simply disable the optimization completely, i.e. `quarkus.arc.remove-unused-beans=false`. Anyway, do we plan to encourage users to depend on Aesh API directly? I can image some kind of abstraction that would work for most common use cases:
```java
public class CapitalizeCommand {
   
   @Inject
   Capitalizer capitalizer;

   @Inject
   @Arguments  
   List<String> arguments;
   
   @Inject
   @Option 
   String file;

   public void execute(@Observes StartupEvent event) {
      Path out = Paths.get(file);
      try {
         Files.write(out, arguments.stream().map(capitalizer::perform).collect(Collectors.toList()), Charset.defaultCharset());
       } catch (IOException e) {
          throw new CommandException(e);
      }
   }
}
```

**Comment by geoand** [comment]
The idea that @stalep had was to be provide the `QuarkusContext`  through the `CommandInvocation` which adds some unique possibilities to access parts of quarkus without exposing any other api.

But even if we don't end up doing that for Aesh, doesn't it make sense to provide a middle-ground for the bean removal optimization?

**Comment by mkouba** [comment]
> But even if we don't end up doing that for Aesh, doesn't it make sense to provide a middle-ground for the bean removal optimization?

The "middle-ground" is in the hands of extensions. I'm not so sure it makes sense to move this functionality in the Arc extension. Not until we have at least two practical use cases we want to support.

**Comment by geoand** [comment]
One reason I wanted to use the `BeanManager` was to have a proxy that I could use to get all instances of a bean and delegate according to some (yet) uspecified condition. I was basically thinking about the testing use case were we want to generate a proxy that could delegate to various alternatives.
I guess that proxy could be created with similar difficulty or ease using `Instance` as with `BeanManager`

**Comment by geoand** [comment]
Closing this since it is probably not useful at the moment. Will reexamine if needed

**Pull request closed without merging**